pkg/rpm/database.go

```diff
@@ -41,9 +41,12 @@ const (
 	// databaseDir is the directory where the RPM database is expected to be in
 	// the container filesystem.
 	databaseDir = "var/lib/rpm"
-	// rhcosDbDir is an alternative to databaseDir, as RHCOS saves its RPM DB in
-	// a different path.
-	rhcosDbDir = "usr/share/rpm"
+
+	// rhcosDBDir is an alternative to databaseDir, as RHCOS saves its RPM DB in
+	// a different path that symlinks to var/lib/rpm.
+	// Since PR #1014, Scanner code follows symlinks , but usr/share/rpm will get filtered out
+	// as it is not a well known database dir unless defined here.
```

Comment:
I spent about an hour trying to understand this. Unrelated thing, but I found a security vulnerability <https://issues.redhat.com/browse/ROX-14035>.

I think what's described here is not how #1014 works. Moreover, #1014 did not actually touch the code which has this effect of skipping `usr/share/rpm`.

The reason why `var/lib/rpm` symlink is not followed is because `node.filesMap` (implementation of `analyzer.Files` interface) is populated initially via `addFiles()` function which uses `filepath.WalkDir()` library function.

https://github.com/stackrox/scanner/blob/3432ef813ebc20a007d185a533bebbb949d41d14/pkg/analyzer/nodes/node.go#L101-L107

In the [documentation](https://pkg.go.dev/path/filepath#WalkDir) of `filepath.WalkDir()` it is said:

> WalkDir does not follow symbolic links.

Basically, `WalkDir` will see that `var/lib/rpm` is a symlink and will not traverse it.

On one hand, `usr/share/rpm` is a viable substitute for now, but in order to move with it long-term, we need to obtain a confirmation that it will stay as a location where CoreOS team will store RPMDB in all future versions.

On the other hand, `var/lib/rpm` is a _convention_ and we'll do better job if we teach our code work with symlinks and respect conventions rather than add more and more special cases.
Unfortunately, dealing with symlinks is difficult and like many other difficult things isn't implemented in Go standard library. It is also additionally complicated by the fact that our symlinks have custom root. If there's no good library solving this problem (I wonder how is it solved in the container scanning!?), I suggest to get a promise from the CoreOS team that they maintain `/usr/share/rpm` as a contract and move on with this.

I'm fine with a TODO for now.

Reply:
Great catch with the potential vuln! If we are to keep our ability to read symlinks, we should definitely handle any known potential vulnerability before releasing. Luckily, we have handled a similar case before when it came to reading ZIP files (see https://security.snyk.io/research/zip-slip-vulnerability for more information about ZIP-slip): https://github.com/stackrox/scanner/blob/2.27.3/pkg/ziputil/open.go#L60

If we keep the symlink stuff, then let's definitely ensure the file is under the `root`. 

Similarly, I have been thinking about just simply looking at `/usr/share/rpm` instead, and preferring that over `/var/lib/rpm` when we know we are looking at RHCOS. Perhaps that could be a good followup after the MVP, so we don't delay this